3주차 Spring Web으로 ToDo REST API 만들기

euijeonghwang opened the pull request:
3주차도 잘 부탁드립니다~

euijeonghwang committed:
HelloController 유닛 테스트해라

HelloController 클래스 생성 및 테스트 코드 작성

euijeonghwang committed:
Task 모델 테스트해라

Task 클래스에 대한 테스트코드 작성

euijeonghwang committed:
TaskService 테스트 작성해라

TaskService에 대한 테스트코드 작성함

euijeonghwang committed:
TaskController 테스트 해라

TaskControllerTest 테스트 코드 작성
TestMethodOrder 애노테이션을 사용하여 Order를 지정하여 테스트를 Order 순서대로 진행하게끔 작성함

euijeonghwang committed:
TaskController MockMVC 테스트 작성해라

MockMVC를 활용한 테스트 코드 작성함

euijeonghwang committed:
HelloController에 대한 MockMvc 테스트 코드를 작성해라

MockMvc를 활용한 HelloController 테스트 코드 작성함

euijeonghwang committed:
Lombok 라이브러리를 추가해라

Lombok 라이브러리를 사용하면 생성자를 만들고, getter/setter 등을 애노테이션으로 편리하게 사용할 수 있습니다. 또한, Build 패턴을 사용하여 객체를 생성할 수 있고, log을 사용해 보안이 취약한 System.out.print와 같은 메서드 대신 로그를 남겨 디버깅이 가능합니다.
이러한 이유로 Lombok을 추가하는 바입니다.

euijeonghwang committed:
Lombok을 사용하여 Task 클래스를 수정해라

get/set 코드를 삭제하고 Getter/Setter 애노테이션을 사용하도록 했습니다.
Builder 애노테이션을 사용하기 위해 NoArgsConstructor와 AllArgsConstructor를 추가하였습니다.

euijeonghwang committed:
MockBean을 주입받고, given 메서드와 verify 메서드를 활용하여 구체적인 테스트를 작성해라

MockBean 애노테이션을 통해 TaskService에 대한 가짜 객체를 생성합니다.
BeforeEach에서 given 메서드를 통해 가짜 객체의 동작에 대해 정의합니다.
verfiy 메서드를 통해 가짜 객체의 메소드가 실행되었는지 체크합니다.

Json을 사용해 Task 객체를 생성합니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -1,23 +1,18 @@
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
Lombok을 사용함으로써 굉장히 코드가 단순해졌네요. Lombok을 사용하면서 장점만 있을까요?

euijeonghwang replied:
제가 생각하는 Lombok을 사용함으로써 얻는 장단점은 아래와 같습니다.

**장점**

- 애노테이션을 사용함으로써 코드가 간결해진다.
- 로그를 확인할 수 있다.

**단점**

- 라이브러리에 의존성이 강해진다.
- Lombok을 사용하기 위한 별도의 설정이 필요하다.
- 컴파일 후의 코드가 달라서 개발자가 의도한대로 실행이 안될 수 있다.

찾아보니 `@NoArgsConstructor`와 `@AllArgsConstructor`를 사용하게 되면 추후 치명적인 버그가 생길 가능성이 있으므로 직접 생성자 코드를 작성하고 그 위에 `@Builder` 애노테이션을 사용함으로써 builder패턴은 유지하도록 하였습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerMockMvcTest.java:
```diff
@@ -0,0 +1,123 @@
+                .willThrow(TaskNotFoundException.class);
+    }
+
+    @Test
+    @DisplayName("Tasks 목록 얻기")
```
테스트 코드에서 무엇을 하고 있는지를 작성해 주신 것 같아요. 이 테스트는 무엇을 테스트하고 있고 무엇을 기대하고 있는지 작성해 주면 좋을 것 같아요.

테스트 코드를 작성하는 것도 중요하지만, 무언가 문제가 생겼을 때 빠르게 알아차리는 것도 중요하거든요

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,77 @@
+    private static final Long TASK_ID = 1L;
+    private final Task task = new Task();
+
+    @Test
+    @Order(0)
```
테스트에 순서가 중요한가봐요. 테스트가 서로 의존하고 있어서 그런걸까요?

euijeonghwang replied:
`@BeforeEach`를 사용하지 않고 테스트 클래스 전체에 `final`로 객체를 선언하고 생성하다보니 각 테스트가 서로 의존하고 있습니다. 따라서 `@Order`를 사용하여 순서대로 테스트를 실행하도록 했는데, 이 경우 하나의 테스트 코드가 변경된다면 도미노처럼 다른 테스트 코드에 영향을 줄 수 있기 때문에 각 테스트 코드가 독립적으로 실행될 수 있도록 수정하겠습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,43 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class TaskTest {
```
`Task` 객체에 대해서 아주 꼼꼼하게 테스트를 작성해 주셨네요! 테스트를 우리가 작성한 코드의 예제라고 생각해 봅시다. 테스트가 꼼꼼하게 작성되어 있다면 하나하나의 기능에 대해서는 좋겠지만, 전체적으로 어떻게 사용하는지는 판단하기 어려울 수도 있을 것 같아요. 따라서 이 객체를 어떻게 사용하는지를 간단하게만 작성해 보는 것도 좋을 것 같아요

```java
Task task = new Task();
task.setId(1L);
task.setTitle("책 읽기");

assertThat(task.getId()).isEqualTo(1L);
assertThat(task.getTitle()).isEqualTo("책 읽기");
```

euijeonghwang committed:
Lombok의 NoArgsConstructor, AllArgsConstructor를 사용하지 마라

Lombok의 애노테이션을 사용해서 생성자를 생성할 때 파라미터 순서를 필드 선언 순서에 맞춰서 생성한다. 따라서 코드의 리팩터링없이 개발자가 인지하지 못한 채로 파라미터 순서를 바꿔줬을 때 어떤 오류도 발생하지 않아 추후 버그가 생길 수도 있다.
대신 직접 생성자를 만들고 Builder 애노테이션을 생성자에 선언해주었다.

euijeonghwang committed:
8fdb4c2 간단히 Task 전체에 대한 테스트 코드를 작성해라

Task 객체를 생성하고 간단히 getId와 getTitle을 검증하는 테스트 코드 작성

euijeonghwang commented:
질문있습니다. 2주차에 작성했던 서비스, 컨트롤러 코드로 바꿔도 되나요? 아니면 이번 주에는 테스트코드가 과제이니 테스트 코드만 작성해도 되나요~?

hannut91 commented:
> 질문있습니다. 2주차에 작성했던 서비스, 컨트롤러 코드로 바꿔도 되나요? 아니면 이번 주에는 테스트코드가 과제이니 테스트 코드만 작성해도 되나요~?

바꾸셔도 됩니다 ㅎㅎ

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/HelloController.java:
```diff
@@ -0,0 +1,13 @@
+@RestController
+public class HelloController {
+
+    @RequestMapping("/")
+    public String sayHello() {
+        return "Hello, World!";
+    }
```
강의에서는 그냥 기본으로 있는 것처럼 보이지만 사실 이 API는 Health check를 위한 API로 쓰이기도 합니다. 서버 실행이 완료되어 응답을 받을 수 있는 상태인지 확인할 수 있죠. 참고로만 알아두시면 좋을 것 같습니다

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerMockMvcTest.java:
```diff
@@ -0,0 +1,123 @@
+    void getTaskWithWrongId() throws Exception {
+        mockMvc.perform(get("/tasks/{id}", 100L))
+                .andExpect(
+                        result -> assertThatThrownBy(() -> service.getTask(100L))
+                                .isInstanceOf(TaskNotFoundException.class)
```
여기서는 어떤 것을 검증하려고 하셨나요?

euijeonghwang replied:
존재하지 않는 id로 `GET` 요청을 받을 경우 `TaskNotFoundException`이 발생하는지 검증하려고 했습니다

hannut91 replied:
Controller에서는 사용자가 실제로 API를 호출하는 것처럼 테스트하는게 목적인 것 같아요. 그러다 보니 `service.getTask`가 잘 동작하는지 테스트하는 것은 Controller에서 하는 것은 아닐 것 같다는 생각이 들었어요

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,64 @@
+    @Test
+    void getTask() {
+        Long id = 1L;
+        assertThat(taskService.getTask(id)).isNotNull();
```
id로 task를 얻는 테스트를 작성해 주신 것 같아요. 역시서 만약 주어진 id에 해당하는 task를 찾을 수 없는 경우에는 어떤 일이 일어나나요? 지금은 `정의되어 있지 않은 상태`입니다. 테스트를 작성해서 정의해 주세요

euijeonghwang committed:
Getter, Setter를 제외한 Lombok 애노테이션 제거

Builder와 생성자 애노테이션을 사용함으로써 toString()도 테스트 커버리지에 추가되어 100% 도달하지 못함. 따라서 Getter/Setter 애노테이션만 남기고 생성자를 코드로 직접 추가함.

euijeonghwang committed:
각 클래스의 테스트 목적에 맞는 테스트 코드 작성해라

TaskControllerTest : 테스트 코드가 Order에 의존하지 않도록 수정
TaskControllerMockMvcTest : Controller에서는 사용자가 실제로 API를 호출하는 것처럼 테스트하도록 수정
TaskServiceTest : 존재하지 않는 ID일 경우 TaskNotFoundException을 던지는 테스트 코드를 작성

euijeonghwang committed:
TaskControllerTest-createTask 테스트 코드 자세히 작성

isGreaterThan(1L) -> isEqualTo(2L) 로 특정 id를 비교하도록 수정함
taskController의 detail 메서드를 통해 가져온 title과 테스트 코드에서 생성한 title이 동일한지 비교함

euijeonghwang committed:
과제 2주차 때 생성한 클래스들로 변경 및 보완

과제 2주차 때 작성했던 코드로 대체 및 보완함

euijeonghwang committed:
CreateTask, DeleteTask 수정

특정 id로 검증 대신 createTask한 후의 Task의 id값을 얻어서 검증하도록 함

euijeonghwang committed:
Controller 테스트 보완 및 MockMvcTest 주석 제거

TaskControllerTest 테스트 코드 보완
TaskControllerMockMvcTest 주석 제거

euijeonghwang committed:
TaskControllerTest detail 메서드 실행 시 에러 해결해라

BeforeEach에서 정의해놓은 title과 같은지 검증하고자 했으나,
patchTask, updateTask가 먼저 실행됨에 따라 에러 발생으로 코드 삭제함

hannut91 commented on app/src/main/java/com/codesoom/assignment/TitleNotFoundException.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * RequestBody로 넘어오는 title이 empty일 때 발생하는 런타임 예외
+ */
+public class TitleNotFoundException extends RuntimeException {
```
이 예외가 `RuntimeException`을 상속해야 하는 이유는 무엇인가요?

euijeonghwang replied:
`TitleNotFoundException`은 사용자의 입력값에 따라 발생할 가능성이 있는 예외이기 때문에,
컴파일 당시에는 알지 못하고 실행해야 알 수 있으므로 `RuntimeException`을 상속받도록 했습니다

hannut91 reviewed (commented):
BDD 스타일로 테스트를 한 번 작성해 보세요. 그리고 왜 이렇게 작성하려고 하는건지 고민해 보시면 좋을 것 같습니다.

https://johngrib.github.io/wiki/junit5-nested/

euijeonghwang committed:
TaskControllerTest - BDD 스타일로 작성

Describe-Context-It 패턴으로 테스트 코드 수정함

euijeonghwang commented:
> BDD 스타일로 테스트를 한 번 작성해 보세요. 그리고 왜 이렇게 작성하려고 하는건지 고민해 보시면 좋을 것 같습니다.
> 
> https://johngrib.github.io/wiki/junit5-nested/

우선 `TaskControllerTest`를 BDD 스타일로 수정해보았습니다. `given-when-then` 스타일은 해보았지만 `Describe-Context-It`스타일은 처음이라 맞게 작성한건지 잘 모르겠습니다. 😅
테스트 코드 결과가 계층화되서 이해하기 쉽다는 장점이 있는 건 확실한 것 같습니다. 
다만 Mock 컨트롤러에도 적용해보려고 하니까 안되서 찾아보니 이슈가 있네요. 일단 Mock 테스트는 제외하고 서비스 테스트부터 BDD로 적용해보겠습니다.

euijeonghwang committed:
TaskServiceTest - BDD 스타일로 작성해라

Describe-Context-It 패턴으로 테스트 코드 수정함

euijeonghwang committed:
MockMvc에서 Exception을 테스트해라

Spring MockMvc에서 Exception 검증하는 테스트 코드 작성함

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,171 @@
+        taskList.clear();
+    }
+
+    @Nested
+    @DisplayName("getTasks 메소드는")
+    class Describe_getTasks {
+        @Nested
+        @DisplayName("Task가 존재한다면")
+        class Context_with_Task {
+            @Test
+            @DisplayName("List의 모든 Task들을 리턴한다")
+            void it_return_tasks() {
+                Task task = new Task();
+                task.setTitle("test");
+                taskService.createTask(task);
+
+                assertThat(taskService.getTasks()).isNotEmpty();
+            }
+        }
+
+        @Nested
+        @DisplayName("Task가 존재하지 않는다면")
+        class Context_with_NoTask {
+            @Test
+            @DisplayName("빈 배열을 리턴한다")
+            void it_return_empty_list() {
+                assertThat(taskService.getTasks()).isEmpty();
+            }
+        }
+    }
```
```java
    @Nested
    @DisplayName("getTasks 메소드는")
    class Describe_getTasks {
        @Nested
        @DisplayName("Task가 존재한다면")
        class Context_with_Task {
            @BeforeEach
            void setUp() {
                Task task = new Task();
                task.setTitle("test");
                taskService.createTask(task);
            }
            
            @Test
            @DisplayName("List의 모든 Task들을 리턴한다")
            void it_return_tasks() {
                assertThat(taskService.getTasks()).isNotEmpty();
            }
        }

        @Nested
        @DisplayName("Task가 존재하지 않는다면")
        class Context_with_NoTask {
            @Test
            @DisplayName("빈 배열을 리턴한다")
            void it_return_empty_list() {
                assertThat(taskService.getTasks()).isEmpty();
            }
        }
    }
```

`beforeEach`로 주어진 상황이 어떻게 달라지는지 표현할 수도 있습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,171 @@
+   @Nested
+   @DisplayName("createTask 메소드는")
+   class Describe_createTask {
+        @Nested
+        @DisplayName("task 객체가 정상적으로 넘어올 경우")
+        class Context_with_task {
```
context는 주어진 상황(given)이 달라질 때를 명시할 수 있으므로, 이 context가 있으면 다른 상황이 있는지 생각하게 돼요. 따라서 다른 상황이 없다면 안써줘도 되고, 있다면 써주는게 좋을 것 같습니다.

euijeonghwang replied:
생각해보니 `Controller`에서 정상적이지 않은 `Task`를 받을 때 `TitleNotFoundException`을 던지므로 해당 `Context`는 삭제하겠습니다

euijeonghwang commented:
질문있습니다 ~

> [b8eaeb5](https://github.com/CodeSoom/spring-week3-assignment-1/pull/91/commits/b8eaeb5430de1e3f77e188a64d0ef32fb416236f) Controller에서는 사용자가 실제로 API를 호출하는 것처럼 테스트하는게 목적인 것 같아요. 그러다 보니 service.getTask가 잘 동작하는지 테스트하는 것은 Controller에서 하는 것은 아닐 것 같다는 생각이 들었어요

과제 풀이 1-1 영상 19분에 아샬님이 작성하신 코드
```java
assertThatThrownBy() -> controller.detail(100L)).isInstanceOf(TaskNotFoundException.class);
```
이 부분은 `service.detail`이 아니라 `controller.detail`이라서 Controller 테스트 클래스에서 작성한게 맞을까요?

euijeonghwang committed:
테스트 코드 보완해라

과제 풀이 영상 보고 테스트 코드 보완함

hannut91 commented:
> 이 부분은 service.detail이 아니라 controller.detail이라서 Controller 테스트 클래스에서 작성한게 맞을까요?

네 ㅎㅎ 저건 MockMVC로 테스트를 작성한게 아니라서 그럴거예요
